libnopegl/src/glsl/helper_noise.glsl

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright 2023 Matthieu Bouron <matthieu.bouron@gmail.com>
+ * Copyright 2023 Clément Bœsch <u pkh.me>
+ * Copyright 2023 Nope Foundry
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#include helper_misc_utils.glsl
+
+float u32tof32(highp uint x)
+{
+    uint y = (0x7FU << 23U) | (x >> 9U);
+    return (uintBitsToFloat(y) - 1.0);
+}
+
+highp uint lowbias32(highp uint x)
+{
+    x ^= x >> 16;
+    x *= 0x7feb352dU;
+    x ^= x >> 15;
+    x *= 0x846ca68bU;
+    x ^= x >> 16;
+    return x;
+}
+
+uint hash(uvec2 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y));
+}
+
+uint hash(uvec3 x)
+{
+    return lowbias32(x.x ^ lowbias32(x.y ^ lowbias32(x.z)));
+}
+
+float random(uvec3 x)
+{
+    return u32tof32(hash(x));
+}
+
+float random_grad(uvec2 x, vec2 f)
+{
+    float r = u32tof32(hash(x)) * 2.0 * ngli_pi;
+    return dot(vec2(cos(r), sin(r)), f);
+}
+
+float random_grad(uvec3 x, vec3 f)
+{
+    uint hash = hash(x) & 15U;
+    float u = hash < 8U ? f.x : f.y;
+    float vt = ((hash == 12U) || (hash == 14U)) ? f.x : f.z;
+    float v = hash < 4U ? f.y : vt;
+    return ((hash & 1U) != 0U ? -u : u) + ((hash & 2U) != 0U ? -v : v);
+}
```

Comment:
I will need explanations for this function at least. It looks like some kind of selector, but a bit unclear.

Also, maybe using only masks rather than mixing masking + sign comparison + direct equalities would help understanding.

Reply:
This is the formula used by Perlin. I added some non-formated comments and a link to the explanation. I will properly reformat the comment / explanation. I'll probably propose another approach based on cos/sin to avoid the whole branching stuff.

Reply:
I've pushed a commit with an alternative grad function.

Reply:
A video "comparing" the two: https://0x5c.me/noise_grad_v2.mp4 (the darker one is the one using the grad function from Perlin).

Reply:
I would keep the perlin function though, this is what people expect when using a noise called  "perlin".

Reply:
After an offline discussion, we decided to keep the trigo version.